In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

# Сингулярное разложение матриц <a class='anchor' id='svd'>

### Задание 1 <a class='anchor' id='svd_1'>

*Найдите посредством NumPy SVD для матрицы:*

$$\begin{pmatrix}
1 & 2 & 0 \\ 
0 & 0 & 5 \\ 
3 & -4 & 2 \\ 
1 & 6 & 5 \\ 
0 & 1 & 0
\end{pmatrix}.$$

In [2]:
A = np.array([[1, 2, 0], [0, 0, 5], [3, -4, 2], [1, 6, 5], [0, 1, 0]])
U, s, W = np.linalg.svd(A)

V = W.T

D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

print(f'Матрица A =\n{A}\n')
print(f'Матрица U =\n{U}\n')
print(f'Матрица D =\n{D}\n')
print(f'Матрица V =\n{V}\n')

print(f'UDV^T =\n{U @ D @ V.T}\n')

print(f'Сингулярные числа =\n{s}')

Матрица A =
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]

Матрица U =
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]

Матрица D =
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]

Матрица V =
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]

UDV^T =
[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1. -0.]]

Сингулярные числа =
[8.82 6.14 2.53]


### Задание 2 <a class='anchor' id='svd_2'>

*Для матрицы из предыдущего задания найдите:*

*а) евклидову норму;*
    
*б) норму Фробениуса.*

Если для матрицы известно сингулярное разложение, то для вычисления евклидовой нормы и нормы Фробениуса можно использовать сингулярные числа. Для матрицы из предыдущего задания они равны

$$\mu_1 = 8.82,$$

$$\mu_2 = 6.14,$$

$$\mu_3 = 2.53.$$

а) евклидову норму

Евклидова норма равна максимальному сингулярному числу, т.е

$$\left \| A \right \|_E = \mu_1 = 8.82.$$

**Ответ: 8.82.**

б) норму Фробениуса

Норма Фробениуса через сингулярные числа вычисляется как

$$\left \| A \right \|_F = \sqrt{\sum_{k = 1}^r \mu_k^2} = 11.04.$$

**Ответ: 11.04.**

In [3]:
# Через сингулярные числа
print(f'Евклидова норма = {np.max(s)}')
print(f'Норма Фробениуса = {np.linalg.norm(s)}\n')

# Проверка
print(f'Евклидова норма numpy = {np.linalg.norm(A, ord=2)}')
print(f"Норма Фробениуса numpy = {np.linalg.norm(A, ord='fro')}")

Евклидова норма = 8.824868854820444
Норма Фробениуса = 11.045361017187263

Евклидова норма numpy = 8.824868854820444
Норма Фробениуса numpy = 11.045361017187261
